compiler/semantic/op.go

```diff
@@ -418,13 +418,20 @@ func semOp(ctx context.Context, scope *Scope, o ast.Op, ds *data.Source, head *l
 			NullsFirst: o.NullsFirst,
 		}, nil
 	case *ast.Head:
-		limit := o.Count
-		if limit == 0 {
-			limit = 1
+		expr, err := semExpr(scope, o.Count)
+		if err != nil {
+			return nil, fmt.Errorf("head: %w", err)
+		}
+		val, err := kernel.EvalAtCompileTime(scope.zctx, expr)
+		if err != nil {
+			return nil, fmt.Errorf("head: %w", err)
+		}
+		if val.AsInt() < 1 {
+			return nil, errors.New("head: expression must be of positive value")
```

Comment:
After testing this branch some, I think we ought to tune the error message a bit, both to make clear that the value must be a positive integer and to show what the actual value is.
```suggestion
			return nil, fmt.Errorf("head: expression value is not a positive integer: %s", zson.MustFormatValue(val))
```